mloop/interfaces.py

```diff
@@ -1,320 +1,329 @@
+        
+    Keyword Args:
+        test_landscape (Optional [TestLandscape]): Landscape that can be given a set of parameters and a cost and other values. If None creates a the default landscape. Default None 
+        out_queue_wait (Optional [float]): Time in seconds to wait for queue before checking end flag.
+    
+    '''
+    def __init__(self, 
+                 test_landscape=None,
+                 **kwargs):
+        
+        super(TestInterface,self).__init__(**kwargs)
+        if test_landscape is None:
+            self.test_landscape = mlt.TestLandscape()
+        else:
+            self.test_landscape = test_landscape
+        self.test_count = 0
+    
+    def get_next_cost_dict(self, params_dict):
+        '''
+        Test implementation. Gets the next cost from the test_landscape.
+        '''
+        
+        self.test_count +=1
+        self.log.debug('Test interface evaluating cost. Num:' + repr(self.test_count))
+        try:
+            params = params_dict['params']
+        except KeyError as e:
+            self.log.error('You are missing ' + repr(e.args[0]) + ' from the in params dict you provided through the queue.')
+            raise
+        cost_dict = self.test_landscape.get_cost_dict(params)
+        return cost_dict
+
+      
+class ShellInterface(Interface):
+    '''
+    Interface for running programs from the shell.
+    
+    Args:
+        params_out_queue (queue): Queue for parameters to next be run by experiment.
+        costs_in_queue (queue): Queue for costs (and other details) that have been returned by experiment.
+        
+    Keyword Args:
+        command (Optional [string]): The command used to run the experiment. Default './run_exp'
+        params_args_type (Optional [string]): The style used to pass parameters. Can be 'direct' or 'named'. If 'direct' it is assumed the parameters are fed directly to the program. For example if I wanted to run the parameters [7,5,9] with the command './run_exp' I would use the syntax::
+            
+                ./run_exp 7 5 9
+            
+            'named' on the other hand requires an option for each parameter. The options should be name --param1, --param2 etc. The same example as before would be ::
+        
+                ./run_exp --param1 7 --param2 5 --param3 9
+            
+            Default 'direct'.
+    '''
+    
+    def __init__(self,
+                 command = './run_exp',
+                 params_args_type = 'direct',
+                 **kwargs):
+        
+        super(ShellInterface,self).__init__(**kwargs)
+        
+        #User defined variables
+        self.command = str(command)
+        if params_args_type == 'direct' or params_args_type ==  'named':
+            self.params_args_type = str(params_args_type)
+        else:
+            self.log.error('params_args_type not recognized: ' + repr(params_args_type))
+        
+        if 'param_names' in kwargs:

```

Comment:
Could you add `param_names` as a named keyword arg (defaulting to `None`) instead of getting it out of `kwargs`, and add a section to the docstring for it? That will get it showing up in the generated docs and whatnot.

Reply:
I can, and that's what I originally did - but entering it as a named keyword arg removes it from the kwargs dict in the superclass constructor that's called. This has the consequence of not writing the param_names list to the archive (as it only writes the keyword args that aren't named in the constructors), so you can't go back to a dataset and see which parameter names you used. The method used was my workaround, do you know a way to satisfy both requirements?

Reply:
I have realised it's a very simple fix. Will change and commit everything. Thanks